ath79: restore userspace support for LEDs AR8327/AR8337 switches

Leo-PL opened the pull request:
At start, TP-Link WDR4300, WR1043NDv2/3 and Archer C7v[1-3] families are converted, as an example for other devices.
I decided to post this by @DittelHome's request, because qca8k support in ath79 still requires a lot of effort, let alone LED supoort, which isn't agreed upon in upstream.

Compile and run-tested on devices mentioned above.

Leo-PL committed:
kernel: ar8327: support LED device tree bindings

The ar8216 switch driver supports exposing configuration of AR8327 and
AR8337 switch LEDs to the userspace, however it is only configurable
through platform data, causing the devices ported from ar71xx target to
lack the support.
Since there is still a long way to go until we can migrate the target to
qca8k, an interim solution is needed.
Extend ar8327_hw_config_of function to parse a "leds"
subnode, which will populate the missing platform data based on device
tree contents, and restore the existing support for the LEDs.
Standard bindings apply, mapping "reg" property to LED index, with
addition of "qca,led-mode" property, which selects HW (0) or SW (1)
mode, defaulting to HW mode.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/12487
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Leo-PL committed:
ath79: support switch LEDs on TL-WDR4300 family

Add switch LED definitions for TP-Link TL-WDR4300 family, based on data
extracted from ar71xx board file. Update the LED labels to match current
pattern, i.e. drop the "tp-link:" prefix.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/12487
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Leo-PL committed:
ath79: support switch LEDs on Archer C7 family

Add switch LED definitions for TP-Link Archer C7 v1/2/3 family, based on data
extracted from ar71xx board file. Update the LED labels to match current
pattern, i.e. drop the "tp-link:" prefix.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/12487
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Leo-PL committed:
ath79: support switch LEDs on TL-WR1043ND v2/v3

Add switch LED definitions for TP-Link TL-WR1043ND family, based on data
extracted from ar71xx board file. Update the LED labels to match current
pattern, i.e. drop the "tp-link:" prefix.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/12487
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Leo-PL commented:
@DittelHome please check if it works for you.

DittelHome commented:
@Leo-PL 
Test was succesful !
Now all leds are available and working.
 
```
root@OpenWrt:/# ls /sys/class/leds
ath10k-phy0   green:lan2    green:qss     green:usb2    green:wlan5g
ath9k-phy1    green:lan3    green:system  green:wan
green:lan1    green:lan4    green:usb1    green:wlan2g
```

neheb commented:
Seems like a good stop gap solution until this can properly be handled upstream with qca8k, which AFAIK is still slower than the swconfig driver.

hauke commented:
qca8k in Linux 6.4 has LED support: 
https://github.com/torvalds/linux/commit/91acadcc6e599dfc62717abcdad58a459cfb1684
https://github.com/torvalds/linux/commit/1e264f9d2918b5737023c44a23ae04def1095210

Leo-PL commented:
> qca8k in Linux 6.4 has LED support: [torvalds/linux@91acadc](https://github.com/torvalds/linux/commit/91acadcc6e599dfc62717abcdad58a459cfb1684) [torvalds/linux@1e264f9](https://github.com/torvalds/linux/commit/1e264f9d2918b5737023c44a23ae04def1095210)
> 
> @Ansuel plans to backport this to kernel 5.15 soon.

Oh, I have missed that - glad to hear, that it's merged.

neheb commented:
ping @Ansuel

Ansuel commented:
@neheb the perf problem still needs to be evaluated...

Leo-PL commented:
Then maybe until qca8k lands in ath79, we might try this. This was whole idea.

neheb commented:
Wonder if a temporary alternative is to write values with mdio.

Leo-PL commented:
@Ansuel TL;DR is the patch adds a way to instantiate LEDs using device tree bindings in ar8237.c The code handling the accesses is untouched. Next patches restore the LEDs with names compatible to ones present in ar71xx. Migration scritps would strip the `tp-link:` prefix from them already.

DittelHome commented:
Today i have tested the OpenWrt Release 23.05.0 in the hope that the WAN LED is working ..
Unfortunately still no WAN LED

hauke commented on target/linux/generic/files/drivers/net/phy/ar8327.c:
```diff
@@ -643,6 +644,38 @@ ar8327_hw_config_of(struct ar8xxx_priv *priv, struct device_node *np)
 		}
 	}
 
+	leds = of_get_child_by_name(np, "leds");
```
You should drop the node again:
```
of_node_put(leds);
```

Leo-PL replied:
Done.

hauke commented on target/linux/generic/files/drivers/net/phy/ar8327.c:
```diff
@@ -643,6 +644,38 @@ ar8327_hw_config_of(struct ar8xxx_priv *priv, struct device_node *np)
+	if (!data->leds)
+		return -ENOMEM;
+
+	for_each_child_of_node(leds, child) {
```
```suggestion
	for_each_available_child_of_node(leds, child) {
```

Leo-PL replied:
Fixed.

hauke commented on target/linux/generic/files/drivers/net/phy/ar8327.c:
```diff
@@ -643,6 +644,38 @@ ar8327_hw_config_of(struct ar8xxx_priv *priv, struct device_node *np)
+	if (!leds)
+		return 0;
+
+	data->leds = kzalloc(of_get_child_count(leds) * sizeof(void *),
```
```suggestion
	data->leds = kvcalloc(of_get_child_count(leds), sizeof(void *),
```

Leo-PL replied:
Fixed.

Leo-PL commented:
Thanks for review - I'll apply suggestions and test them out shortly. I'll review the bindings format as well, to get a closer match to qca8k.

Leo-PL commented:
@hauke I finally had the time to test out the changes - for now only the minimum ones. Still thinking about the bindings - I'm not sure if it's worth to mimic in swconfig what's done in qca8k, by introducing ports hierarchy - but it certainly can be done.
@Ansuel am I right that `default-state = "keep"` will keep the LEDs in HW driven mode when set in qca8k?
If so, I could drop that custom `qca,led-mode` thing that originated from very early qca8k versions.

DittelHome commented:
Maybe we can get the patch out before the official 24.10 release...

Leo-PL commented:
Setting to draft, as I need to update the driver to properly use new `color` and `function` properties for the LEDs.

Leo-PL commented:
@hauke @Ansuel updated to support new style color/function bindings. Ready once again.

Leo-PL commented:
I found some more strange things in the LED support, so setting as draft again.

Leo-PL committed:
kernel: ar8327: fix active-low LED initialization

Switch LEDs configured as active-low remain low instead of high upon
initialization, because in ar8327_leds_init, no distinction is made with
regards to LED pattern based on active_low property - only whether HW
mode is active. Select the proper initial pattern based also on
active_low to fix that.

While at that, simplify the equation ruling pattern selection for
setting brightness, avoiding unnecessary binary XOR operation, not
really valid for 'bool' type.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/12487
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Leo-PL committed:
ath79: TP-link TL-WDR4300: use fixed-link for AR8327 switch connection

Attaching PHY driver to the switch, while adding LEDs binding causes the
PHY driver to create additional LED instances, handled incorrectly by
the PHY driver, which are non-functional. Use fixed-link to attach the
switch driver, instead of PHY driver, to prevent that.

This has a side effect of not logging switch port up/down events in the kernel
log.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/12487
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Leo-PL committed:
ath79: TP-link Archer C7v2: use fixed-link for AR8327 switch connection

Attaching PHY driver to the switch, while adding LEDs binding causes the
PHY driver to create additional LED instances, handled incorrectly by
the PHY driver, which are non-functional. Use fixed-link to attach the
switch driver, instead of PHY driver, to prevent that.

This has a side effect of not logging switch port up/down events in the kernel
log.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/12487
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Leo-PL committed:
ath79: TP-link TL-WR1043ND v2/v3: use fixed-link for AR8327 switch connection

Attaching PHY driver to the switch, while adding LEDs binding causes the
PHY driver to create additional LED instances, handled incorrectly by
the PHY driver, which are non-functional. Use fixed-link to attach the
switch driver, instead of PHY driver, to prevent that.

This has a side effect of not logging switch port up/down events in the kernel
log.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/12487
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

RadxaYuntian commented:
Can we convert switch to DSA first before adding more stuff to the switch? I took out my Archer C7 today and just realize that it was still using the old switch driver.

Leo-PL commented:
@RadxaYuntian There is a significant performance penalty from DSA on this target, unless someone implements hardawre flow offload. If it was that easy, it would be done by now.

RadxaYuntian commented:
I assume the DSA driver is from the upstream, while the legacy driver is from [here](https://github.com/openwrt/openwrt/blob/main/target/linux/generic/files/drivers/net/phy/ar8327.c)?

neheb commented:
is it really that significant?

Leo-PL commented:
> I assume the DSA driver is from the upstream, while the legacy driver is from [here](https://github.com/openwrt/openwrt/blob/main/target/linux/generic/files/drivers/net/phy/ar8327.c)?

What point are you trying to make here?

BTW. This PR is also pre-requisite to #17617, to get the LEDs supported, as Mikrotik engineers in their wisdom chose to connect all controllable LEDs on that unit to AR8327 LED outputs.

Leo-PL commented:
Not going to happen any time soon, unless someone has the knowledge and a lot of free time on their hands.
And this is irrelevant to this PR.

Leo-PL commented:
Marking as ready again, as I've found the root cause - the generic PHY driver was registering non-functional LED entries, which conflicted with ones registered by swconfig.

hauke commented:
Please add the suggestions I made in November again.

Leo-PL commented:
@hauke added. I wonder how I've lost them during the rebases when fixing other stuff.

DittelHome commented:
Thanks to everyone for the fix !!
Is it possible to commit it also to the 24.10 branch ?

DittelHome commented:
@ Leo-PL  OK, three weeks later .., and just before the 24.10.1 release. Maybe it is a good time for backporting ...

Leo-PL commented:
I didn't see complains yet, and I myself would like to have it in official builds - so I'll take a look.

DittelHome commented:
@Leo-PL
Push
Just a reminder ..
